packages/contracts/contracts/core/ManagedService.sol

```diff
@@ -0,0 +1,117 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import { AccessControl } from "@openzeppelin/contracts/access/AccessControl.sol";
+import { console } from "sphinx-forge-std/console.sol";
+
+/**
+ * @title ManagedService
+ * @notice Contract controlled by the Sphinx managed service. This contract is used by
+           the managed service to remotely execute deployments.
+           Users can opt in to this functionality if they choose to do so.
+ */
+contract ManagedService is AccessControl {
+    /**
+     * @dev Role required to make calls through this contract.
+     */
+    bytes32 public constant RELAYER_ROLE = keccak256("RELAYER_ROLE");
+
+    /**
+     * @notice Emitted when funds are withdrawn from this contract.
+     *
+     * @param recipient The recipient that withdrew the funds.
+     * @param amount    Amount of ETH withdrawn.
+     */
+    event Withdew(address indexed recipient, uint256 amount);
+
+    /**
+     * @notice Emitted when a call is made to a remote contract.
+     *
+     * @param relayer The address of the account that made the call.
+     * @param to      The address of the remote contract.
+     * @param data    A keccak256 hash of the input data.
+     */
+    event Called(address indexed relayer, address indexed to, bytes32 data);
+
+    /**
+     * @notice A modifer that refunds the caller for the gas spent in the function call.
+     */
+    modifier refund {
+        uint256 start = gasleft();
+        _;
+        uint256 spent = start - gasleft() + 40000;
```

Comment:
How likely is it that at some point, the `ManagedService` won't have enough ETH in its balance to cover a gas amount equal to `start - gasleft() + 40000`? This'd cause the entire tx to revert, including the user's deployment, which seems potentially problematic

Reply:
If we run out of funds, then we won't be able to execute their deployment regardless of if we revert here or not. This just makes it an explicit requirement that the managed service must contain the necessary funds for each transaction instead of having it be an implicit requirement that the individual wallets must have funds.